clang/test/CodeGenSYCL/invoke-function-addrspace.cpp

```diff
@@ -0,0 +1,31 @@
+// RUN: %clang_cc1 -fsycl -sycl-std=2020 -fsycl-is-device -fsycl-allow-func-ptr -internal-isystem %S/Inputs -disable-llvm-passes -triple spir64-unknown-unknown-sycldevice -emit-llvm -o - %s | FileCheck %s
+
+// Test that the type of function object invoked from the kernel has
+// the right address space.
+
+#include "sycl.hpp"
+
+using namespace cl::sycl;
+queue q;
+
+// CHECK-DAG: define dso_local spir_func i32 @{{.*}}bar10{{.*}}()
+int bar10() { return 10; }
+
+// CHECK-DAG: define linkonce_odr spir_func i32 @{{.*}}invoke_function{{.*}}(i32 () addrspace(4)* %f)
+template <typename Callable>
+auto invoke_function(Callable &&f) {
+  // CHECK-DAG: %f.addr = alloca i32 () addrspace(4)*, align 8
+  // CHECK-DAG: %f.addr.ascast = addrspacecast i32 () addrspace(4)** %f.addr to i32 () addrspace(4)* addrspace(4)*
```

Comment:
Can you please explain the address space cast happening here ?

Reply:
The function object being passed in is allocated in address space 4.  The function bar10 is created in the generic address space.   When bar10 is being converted to 'f', there is an address space mismatch that gets complained about.   This cast is to avoid that.   This is the CodeGen change.

Reply:
I am so sorry I missed your review questions before.

Reply:
This is the way CodeGen aligns object allocation (in address space 0 aka `private`) with object references (in address space 4 aka `generic`).
A little bit more details here - https://github.com/intel/llvm/pull/2864.

Reply:
ok. Thanks for explanation!